services/mfa/mfa.go

```diff
@@ -51,6 +54,69 @@ func getIssuerFromUrl(uri string) string {
 	return url.QueryEscape(issuer)
 }
 
+func generateCode(n int) string {
+	const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = charset[rand.Int63()%int64(len(charset))]
+	}
+	return string(b)
+}
+
+func (m *Mfa) GenerateRecovery(user *model.User) ([]string, *model.AppError) {
+	if err := m.checkConfig(); err != nil {
+		return nil, err
+	}
+
+	rand.Seed(time.Now().UnixNano())
+	codes := make([]string, 5)
+	for i := range codes {
+		codes[i] = generateCode(10)
+	}
+	jsonCodes, _ := json.Marshal(codes)
+
+	if err := m.Store.User().UpdateMfaRecovery(user.Id, string(jsonCodes)); err != nil {
+		return nil, model.NewAppError("GenerateRecovery", "mfa.generate_recovery.save_codes", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	return codes, nil
+}
+
+func (m *Mfa) LoginWithRecovery(user *model.User, code string) *model.AppError {
```

Comment:
Instead of having `LoginWithRecovery`, I'd only check here if the code passed in the second parameter is in the `MfaRecovery` struct field. The function is just to check the recovery code, so something along `CheckRecovery` is easier to grasp when using it.

Reply:
`CheckRecovery` might be misleading, since the function does not only check the code, it also removes it, so it cannot be used again.
Maybe something along the lines of `UseRecoveryCode`?

Reply:
Not a problem, at all :) Changed now.